torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -17,17 +23,190 @@ namespace std {
+
+  void barrier(THDGroup group_id = THDGroupWORLD) override;
+
+  THDGroup newGroup(const std::vector<rank_type>& ranks) override;
+
+private:
+  template<typename T>
+  void allGatherT(std::vector<thpp::Tensor*>& output,
+                  thpp::Tensor& input, THDGroup group_id);
+
+  template<typename T>
+  void allReduceT(thpp::Tensor& data, THDReduceOp operation,
+                  THDGroup group_id = THDGroupWORLD);
+
+  template<typename T>
+  void broadcastT(thpp::Tensor& data, rank_type src_rank,
+                  THDGroup group_id = THDGroupWORLD);
+
+  rank_type _rank; // Current process' rank
+  std::string _addr;
+  port_type _port;
+  rank_type _num_processes; // Number of processes in network
+  std::shared_ptr<store_type> _store;
+  std::shared_ptr<::gloo::transport::Device> _device;
+  std::unordered_map<THDGroup, DataChannel::Group> _groups;
+  
+  std::unique_ptr<GlooCache> _cache;
+
+  // Workers
+  QueueWorker _send_worker, _receive_worker;
+};
+
+
+template<DataOperation D, typename T>
+struct algorithm_spec;
+
+struct GlooCache {
+  using buffer_type = char;
+  using algorithm_type = ::gloo::Algorithm;
+  using context_type = ::gloo::rendezvous::Context;
+  using store_type = ::gloo::rendezvous::PrefixStore;
+
+  using key_type = std::tuple<
+    DataOperation, // operation
+    THDGroup,      // group
+    std::size_t,   // input buffer bytes
+    std::size_t,   // output buffer bytes
+    THDReduceOp,   // reduce op
+    rank_type      // src/dest rank
+  >;
+  using value_type = std::tuple<
+    std::shared_ptr<algorithm_type>, // algorithm
+    std::shared_ptr<buffer_type>,    // input buffer (nullptr if not used)
+    std::shared_ptr<buffer_type>     // output buffer (nullptr if not used)
+  >;
+
+  GlooCache(rank_type rank, std::shared_ptr<::gloo::transport::Device> device,
+            std::shared_ptr<DataChannelGloo::store_type> store)
+   : _rank(rank)
+   , _device(device)
+   , _store(store)
+  {}
+
+  GlooCache(GlooCache const&)      = delete;
+  void operator=(GlooCache const&) = delete;
+
+  std::shared_ptr<context_type> createContext(
+    const DataChannel::Group& group,
+    store_type& store
+  ) {
+    auto context = std::make_shared<context_type>(group.mustGetGroupRank(_rank), group.size());
+    context->connectFullMesh(store, _device);
+    return context;
+  }
+
+  std::shared_ptr<buffer_type> createBuffer(std::size_t bytes) const {
+    return std::shared_ptr<buffer_type>(new char[bytes],
+                                        std::default_delete<char[]>());
+  }
+
+  template<DataOperation D, typename T, typename... Args>
+  value_type getAlgorithm(THDGroup group_id, const DataChannel::Group& group,
+                          Args... args) {
+    auto key = algorithm_spec<D, T>::key(group_id, args...);
+    if (_algorithms.find(key) == _algorithms.end()) {
+      // create prefix store with unique prefix
+      store_type prefix_store(print_key(key), *_store);
```

Comment:
Sorry, cannot see that :C

-> If those two threads are in the same process group and create allReduce ops with the same key at the same time that means that they are performing same algorithm so this is what we want
-> If those two threads are in different process group - this situation cannot happen
-> If those two threads operate on the same `DataChannelGloo` - not sure how can this happen